src/distributions/bernoulli.rs

```diff
@@ -0,0 +1,82 @@
+//! The Bernoulli distribution.
+
+use Rng;
+use distributions::Distribution;
+
+/// The Bernoulli distribution.
+///
+/// This is a special case of the Binomial distribution where `n = 1`.
+///
+/// # Example
+///
+/// ```rust
+/// use rand::distributions::{Bernoulli, Distribution};
+///
+/// let d = Bernoulli::new(0.3);
+/// let v = d.sample(&mut rand::thread_rng());
+/// println!("{} is from a Bernoulli distribution", v);
+/// ```
+#[derive(Clone, Copy, Debug)]
+pub struct Bernoulli {
+    /// Probability of success.
+    p: f64,
+}
+
+impl Bernoulli {
+    /// Construct a new `Bernoulli` with the given probability of success `p`.
+    ///
+    /// # Panics
+    ///
+    /// If `p < 0` or `p > 1`.
+    #[inline]
+    pub fn new(p: f64) -> Bernoulli {
+        assert!(p >= 0.0, "Bernoulli::new called with p < 0");
+        assert!(p <= 1.0, "Bernoulli::new called with p > 1");
+        Bernoulli { p }
+    }
+}
+
+impl Distribution<bool> for Bernoulli {
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> bool {
+        rng.gen_bool(self.p)
```

Comment:
The most expensive piece here is multiplying `p` and converting it to an integer, and a little the bounds test. Can you instead move the initialization to `new`. Than we have a real reason to use a distribution here.

Also I reopened the issue because we wanted something with more precision than `gen_bool`. If you multiply by `u64::max` and use `gen::<u64>` you get exactly 64 bits of precision, which seems pretty close to using `HighPrecision01` which @dhardy and I originally wanted to use. And this should be faster, at least when the set-up happens in `new`.

Also, does it make sense to make this version exact?
With the multiply method and using `u64`s, there are exactly `2^64 + 1` steps between `0.0` and `1.0`. If we want to sample without any bias (not that that is worth much), always return true for the `1.0` case (possibly without using the RNG), multiply by `u64::MAX + 1`, and compare the results with `<=`. I am curious what the performance will be. Interested in testing?

Reply:
On second thought, the are `EPSILON / 4 * 2^64 + 1` values that round up to 1.0 and can't be represented by `f64`. Given that we are already in the rounding error territory, I see no reason to special-case `1.0`. Multiplying with `u64::MAX` and comparing with `<` should just be good enough.

Reply:
Um, no, `f64` has 53 bits of precision IIRC and `u64::MAX` isn't exactly representable in `f64`; I seem to remember you saying we should use this method with `u64` @pitdicker.

I suppose we could multiply by `2.0.powi(64)` but there may still be a problem converting to `u64` afterwards — does it clamp to its max value if out of bounds?

Reply:
> `u64::MAX` isn't exactly representable in `f64`

Oops, yes. I thought too much about how all representable values for `p` in the range `0.0..1.0` are also representable when multiplied in `0..2^64`, that I forgot that multiplying by 2^64 - 1  rounds to multiplying by 2^64. So this method does not work without special-casing 1.0, per my first comment.

Reply:
> I suppose we could multiply by 2.0.powi(64) but there may still be a problem converting to u64 afterwards — does it clamp to its max value if out of bounds?

No, it's [much worse](https://github.com/rust-lang/rust/issues/10184). If we want more than 32 bits of precision, we should probably use the traditional `rng.gen::<f64>() <= p` method.